### PR TITLE
Swap: Ensure a checkpoint is only set once and in the correct order

### PR DESCRIPTION
**Pull request by TheCharlatan**
This also solves the problem where a checkpoint of the same state would be written to the database again on restore.

### CODE REVIEW

File: src/swapd/swap_state.rs [hunk 1]
```diff
@@ -481,4 +515,60 @@ impl State {
...
+        if let State::Bob(BobState::RevealB { checkpoint, .. }) = self {
+            if checkpoint.is_none() {
+                *checkpoint = Some(SwapCheckpointType::CheckpointBobPreLock);
+                true
+            } else {
+                debug!("checkpoint bob pre lock already set");
+                false
+            }
+        } else {
+            error!("Not on CoreArbB state");
```
Comment (TheCharlatan) [inline src/swapd/swap_state.rs:556]: Wrong message.

File: src/swapd/swap_state.rs [hunk 2]
```diff
@@ -75,14 +78,18 @@ pub enum BobState {
...
         cancel_seen: bool,
         remote_params: Params,
         local_params: Params,
         b_address: bitcoin::Address,
     }, // lock (not signed), cancel_seen, remote
     #[display("BuySig")]
-    BuySigB { buy_tx_seen: bool },
+    BuySigB {
+        buy_tx_seen: bool,
+        checkpoint: SwapCheckpointType,
```
Comment (zkao) [inline src/swapd/swap_state.rs:91]: `checkpoint_type`?

Comment (zkao) [reply src/swapd/swap_state.rs:91]: when i read, I assumed the entire checkpoint was in the state

File: src/swapd/runtime.rs [hunk 3]
```diff
@@ -771,22 +771,24 @@ impl Runtime {
...
-                                last_msg: Msg::BuyProcedureSignature(buy_proc_sig.clone()),
-                                enquirer: self.enquirer.clone(),
-                                temporal_safety: self.temporal_safety.clone(),
-                                txs: self.txs.clone(),
-                                txids: self.syncer_state.tasks.txids.clone(),
-                                pending_requests: self.pending_requests.clone(),
-                                xmr_addr_addendum: self.syncer_state.xmr_addr_addendum.clone(),
-                            }),
-                        )?;
+                        if self.state.a_sup_checkpoint_pre_buy() {
```
Comment (zkao) [inline src/swapd/runtime.rs:774]: this one comment extends to the rest of the pr related to state updates, sup. Here one updates the state before being sure that checkpoint_send is successful, therefore generating some potential invalid and unhandled cases

**Review by zkao — APPROVED** [review]
Excellent PR, minor comments related to QA

File: src/swapd/swap_state.rs (hunk 2 above)
Comment (TheCharlatan) [reply src/swapd/swap_state.rs:91]: `last_checkpoint_type` :)

Comment (TheCharlatan) [reply src/swapd/swap_state.rs]: Done in f44aa830eabd6de683354f1cb197b45287943fe3

File: src/swapd/runtime.rs (hunk 3 above)
Comment (TheCharlatan) [reply src/swapd/runtime.rs:774]: Yeah, we need a workflow for critical calls like this where we wait for a response from the database before progressing in the protocol. But I need to think about this some more, so work for another PR.

Comment (zkao) [reply src/swapd/runtime.rs:774]: you're right, thats a larger problem

**Pull request merged**